python/S3_VPC_ENDPOINT_ENABLED/S3_VPC_ENDPOINT_ENABLED.py

```diff
@@ -0,0 +1,399 @@
+
+Reports on:
+   AWS::EC2::VPC
+
+Rule Parameters:
+   None
+
+Scenarios:
+  Scenario 1:
+  Given: No VPC is present while using the describe_vpcs() API call
+   Then: Return NOT_APPLICABLE
+
+  Scenario 2:
+  Given: At least one VPC is present
+    And: The S3 service is not present in the list of "ServiceName" on DescribeVpcEndpoints API
+   Then: Return NON_COMPLIANT on this VPC
+
+  Scenario 3:
+  Given: At least one VPC is present
+    And: The S3 service is present in the "ServiceName" key on DescribeVpcEndpoints API
+    And: The "State" key value is not "Available"
+   Then: Return NON_COMPLIANT on this VPC
+
+  Scenario 4:
+  Given: At least one VPC is present
+    And: The S3 service is present in the "ServiceName" key on DescribeVpcEndpoints API
+    And: The "State" key value is "Available"
+   Then: Return COMPLIANT on this VPC
+'''
+import json
+import sys
+import datetime
+import boto3
+import botocore
+
+
+try:
+    import liblogging
+except ImportError:
+    pass
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::EC2::VPC'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def get_vpcendpoints(vpc_id, event):
+    compliance_results = {}
+    ec2_client = get_client('ec2', event)
+    compliance = 'NON_COMPLIANT'
+    annotate = 'There are no Amazon S3 VPC endpoints present in '+ vpc_id+'.'
+    response = ec2_client.describe_vpc_endpoints(Filters=[
+                {
+                    'Name':'vpc-id',
+                    'Values':[vpc_id]
+                }])
+    for vpce in response['VpcEndpoints']:
+        endpoint_name = vpce['ServiceName']
+        endpoint_state = vpce['State']
+        if is_s3endpoint(endpoint_name):
+            annotate = 'The Amazon S3 VPC endpoint is not in Available state '+vpc_id+'.'
+            if is_available(endpoint_state):
+                annotate = None
+                compliance = 'COMPLIANT'
+    compliance_results['%s' %(vpc_id)] = [annotate, compliance]
+    return compliance_results
+
+
+def is_s3endpoint(endpointname):
+    if 's3' not in endpointname:
+        return False
+    return True
+
+def is_available(endpointstate):
+    if endpointstate == 'available':
+        return True
+    return False
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    vpc_list = []
+    evaluations = []
+    ec2_client = get_client('ec2', event)
+    account_id = event['accountId']
+    vpc_response = ec2_client.describe_vpcs()
+
+    try:
+        if not vpc_response['Vpcs']:
+            evaluations.append(build_evaluation(account_id, 'NOT_APPLICABLE', event))
+        for vpc in vpc_response['Vpcs']:

```

Comment:
if there are no VPCs in the region then `vpc = vpc_response['Vpcs']['VpcId']` will throw an error. The reason I separated these two out to create non_applicable evaluation if there are no VPCs to evaluate in first place. Please let me know if that makes sense.

Reply:
I am talking about L113 and L115 loops. They are doing the same thing, going through the list of vpcs.

```
    for vpc in vpc_response['Vpcs']:
        vpc_list.append(vpc['VpcId'])
    for vpc in vpc_list:
        evaluation_payload = get_vpcendpoints(vpc, event)
        evaluations.append(build_evaluation(vpc, evaluation_payload[vpc][1], event, annotation=evaluation_payload[vpc][0]))
```
is the same than
```
    for vpc in vpc_response['Vpcs']:
        evaluation_payload = get_vpcendpoints(vpc['VpcId'], event)
        evaluations.append(build_evaluation(vpc['VpcId'], evaluation_payload[vpc['VpcId']][1], event, annotation=evaluation_payload[vpc['VpcId']][0]))
```

Reply:
@jongogogo I guess @bksarthak is suggesting the code will give a keyError if VPCs are not present. 
However, @bksarthak since you are already checking if vpc_response contains VPCs or not, so you can combine the loops

if not vpc_response['Vpcs']:
            evaluations.append(build_evaluation(account_id, 'NOT_APPLICABLE', event))
#so if there is no response, you need to return from here and not continue with line 117 and hence avoid the KeyError

Let me know if you have any doubts.

Reply:
ah ok, I understood the issue now. Sorry for the confusion. That makes sense. I was definitely doing two loops. 
Removed the `vpc_list`list and looping through the response now. Added a `return evaluations` within the `if not vpc_response` so that it'll exit the function when there are not VPCs present